src/builtins/fish_key_reader.rs

```diff
@@ -0,0 +1,278 @@
+        }
+    }
+
+    let Some(tail) = recent_keys
+        .len()
+        .checked_sub(4)
+        .and_then(|start| recent_keys.get(start..))
+    else {
+        return false;
+    };
+    let tail = tail.iter().map(|c| c.codepoint);
+    tail.clone().eq("exit".chars()) || tail.eq("quit".chars())
+}
+
+/// Return the name if the recent sequence of characters matches a known terminfo sequence.
+fn sequence_name(recent_chars: &mut Vec<u8>, c: char) -> Option<WString> {
+    if c >= '\u{80}' {
+        // Terminfo sequences are always ASCII.
+        recent_chars.clear();
+        return None;
+    }
+
+    let c = c as u8;
+    recent_chars.push(c);
+    while recent_chars.len() > 8 {
+        recent_chars.remove(0);
+    }
+
+    // The entire sequence needs to match the sequence, or else we would output substrings.
+    input_terminfo_get_name(&str2wcstring(recent_chars))
+}
+
+/// Process the characters we receive as the user presses keys.
+fn process_input(streams: &mut IoStreams, continuous_mode: bool, verbose: bool) -> i32 {
+    let mut first_char_seen = false;
+    let mut queue = InputEventQueue::new(STDIN_FILENO);
+    let mut recent_chars1 = vec![];
+    let mut recent_chars2 = vec![];
+    streams.err.appendln("Press a key:\n");
+
+    terminal_protocols_enable_ifn();
+    while (!first_char_seen || continuous_mode) && !check_exit_loop_maybe_warning(None) {
+        let evt = queue.readch();
+
+        let CharEvent::Key(kevt) = evt else {
+            continue;
+        };
+        let c = kevt.key.codepoint;
+        if c == key::Invalid {
+            continue;
+        }
+        if verbose {
+            streams.out.append(L!("# decoded from: "));
+            for byte in kevt.seq.chars() {
+                streams.out.append(char_to_symbol(byte));
+            }
+            streams.out.append(L!("\n"));
+        }
+        streams
+            .out
+            .append(sprintf!("bind %s 'do something'\n", kevt.key));
+        if let Some(name) = sequence_name(&mut recent_chars1, c) {
+            streams
+                .out
+                .append(sprintf!("bind -k %ls 'do something'\n", name));
+        }
+
+        if continuous_mode && should_exit(streams, &mut recent_chars2, kevt.key) {
+            streams.err.appendln("\nExiting at your request.");
+            break;
+        }
+
+        first_char_seen = true;
+    }
+    0
+}
+
+/// Setup our environment (e.g., tty modes), process key strokes, then reset the environment.
+fn setup_and_process_keys(streams: &mut IoStreams, continuous_mode: bool, verbose: bool) -> i32 {
+    signal_set_handlers(true);
+    // We need to set the shell-modes for ICRNL,
+    // in fish-proper this is done once a command is run.
+    unsafe { libc::tcsetattr(0, TCSANOW, &*shell_modes()) };
+    terminal_protocol_hacks();
+
+    if continuous_mode {
+        streams
+            .err
+            .appendln("To terminate this program type \"exit\" or \"quit\" in this window,");
```

Comment:
not sure if the leading blank line here is dropped intentional

Reply:
I have no idea what it's for, so I dropped it.

Reply:
Okay I put it back